### PR TITLE
doc: cache: Create caching overview

### PR DESCRIPTION
**Pull request by dpw13**
Expands the caching documentation to include a high-level overview of caching strategies and the features available in Zephyr

[Doc preview](https://builds.zephyrproject.io/zephyr/pr/82624/docs/index.html)

### CODE REVIEW

**Comment by dpw13** [comment]
This PR covers a need I've seen on the Discord server to communicate the basics of caching. This is going to be a bit basic for some folks, but I'm hoping it reduces the learning curve for some of the more advanced features in Zephyr.

This is also just my first pass at this documentation. I've seen feedback to break the bulk of the new documentation here into a separate page; I plan to do that in the next few days. Any other feedback on organization or content is welcome.

Currently missing from this PR is alignment with `doc/kernel/usermode/memory_domain.rst` which to be honest I have yet to digest. I hope to devote some time to studying that page and integrating it into this documentation also in the next few days.

File: doc/hardware/cache/index.rst [hunk 1]
```diff
@@ -1,9 +1,146 @@
...
+Globally Disabling the Data Cache
+---------------------------------
+
+As mentioned above, globally disabling data caching can have a significant performance
+impact but can be useful for debugging.
+
+* Requirements:
+  * :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+  * :kconfig:option:`CONFIG_CACHE_MANAGEMENT`: cache API enabled.
+  * Call :cpp:func:`sys_cache_data_disable()` to globally disable the data cache.
```
Comment (kartben) [inline doc/hardware/cache/index.rst:43]:
```suggestion
  * Call :c:func:`sys_cache_data_disable()` to globally disable the data cache.
```

elsewhere too

Comment (dpw13) [reply doc/hardware/cache/index.rst]: Thanks, I'm still a bit new to sphinx. I'm working through building these locally and fixing the errors...

Comment (kartben) [reply doc/hardware/cache/index.rst]: No worries, this one just caught my eye. Let me know if you are having issues setting up the local build environment and if there is anything I can do to help

Comment (dpw13) [reply doc/hardware/cache/index.rst]: It took a bit to figure out how to get `sphinx-build` not to OOM my whole machine, but once that was solved I was able to fix the other errors. Hopefully the docs should now be reviewable in the CI build once it's complete.

File: doc/hardware/cache/index.rst [hunk 2]
```diff
@@ -1,9 +1,166 @@
...
+tools, see :ref:`cache_config` for Zephyr Kconfig options or :ref:`cache_api` for
+the API reference. This section primarily focuses on the data cache though there is
+typically also an instruction cache for systems with cache support.
+
+.. note::
+
+  The information here assumes that the architecture-specific MPU support is enabled.
+  See the architecture-specific documentation for details.
+
+When dealing with multiple bus masters that do not share the same cache, the issue
```
Comment (andyross) [inline doc/hardware/cache/index.rst:17]: This paragraph seems likely to scare people. :)  In fact all but one of our SMP devices enforce inter-CPU cache coherence[1] at the level of the architecture ISA.  The one that doesn't, Xtensa, has a somewhat complicated layer that works very hard to present a vision of memory to app code that is as coherent as possible (it makes shared .data/.bss memory uncached, but leaves stacks cached, and has a validation layer to ensure that nothing the kernel recognizes as shared memory gets used via the wrong pointer).  Apps aren't really expected to have to use this API to manage regular data, it's more for use with DMA frameworks, etc...

Basically my vote would be to punt here in the general docs and just state something like "For inter-CPU shared data on SMP systems, Zephyr in general ensures that memory will be seen in a coherent state.  Most applications will only need to use the cache APIs for interaction with external hardware like DMA controllers or foreign CPUs running a different OS image."  And then maybe point to the CONFIG_KERNEL_COHERENCE kconfig docs (or an elaborated RST page) for the special case of Xtensa.

[1] But not memory ordering, which is a related idea but addressed by the barrier API and not the cache layer.

Comment (dpw13) [reply doc/hardware/cache/index.rst:17]: That seems reasonable to me. My focus was mainly on DMA but I figured I'd throw SMP concerns in as well (though I haven't researched these in Zephyr very well at all). I'll update the docs to punt on the SMP issues as they're handled somewhat differently, as you said.

Comment (dpw13) [reply doc/hardware/cache/index.rst:17]: Added a note about SMP-specific cache coherence and added a link to CONFIG_KERNEL_COHERENCE. I also removed references in the remaining new content to synchronizing shared memory between cores and just focused on DMA and peripherals.

**Review by andyross — DISMISSED** [review]

File: doc/hardware/cache/guide.rst [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+* :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+
+* :kconfig:option:`CONFIG_MEM_ATTR`: enable the ``mem-attr`` library for
+  handling memory attributes in the device tree.
+
+* Annotate your device tree according to :ref:`mem_mgmt_api`.
+
+Assuming the MPU driver is enabled, it will configure the specified regions
+according to the memory attributes specified during kernel initialization. When
```
Comment (erwango) [inline doc/hardware/cache/guide.rst:76]: There's an additional gotcha here in some architectures in which number of MPU regions might not be sufficient if used for other proposes such as userspace. I don't know if that's the right place to discuss this but globally this is another choice that you might have to do in your general application design.

**Review by erwango — DISMISSED** [review]
One comment, but otherwise definitely welcomed anyway. Thanks

File: doc/hardware/cache/guide.rst (hunk 3 above)
Comment (dpw13) [reply doc/hardware/cache/guide.rst:76]: Great point. I've added a couple notes calling this out and pointed to documentation for other features that may also require the use of dedicated MPU regions. Take a look and let me know if my additions match what you're thinking.

Comment (erwango) [reply doc/hardware/cache/guide.rst:76]: Perfect, thanks

**Review by erwango — DISMISSED** [review]

**Review by peter-mitsis — DISMISSED** [review]

File: doc/hardware/cache/guide.rst [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+using a dedicated uncached region of memory, the linker needs to be instructed
+to place buffers into that region. This can be accomplished by specifying the
+memory region explicitly using ``Z_GENERIC_SECTION``:
```
[the remark this answers is not in the document]
Comment (dpw13) [reply doc/hardware/cache/guide.rst:79]: Wow, I had no idea that existed. That seems useful but that capability only seems able to relocate both `.text` and `.data`. Relocating buffers into an uncached region using this mechanism is possible, I suppose, but comes along with all the impact of relocating `.text` to that same region. I feel like `zephyr_code_relocate` is more valuable for placing compiled code that you don't want to modify (third-party DSP code, for example) into a special region than it is for mapping buffers to an uncached region. I'd like to leave references to `zephyr_code_relocate` out of this particular document as I think it'll mostly confuse people. If you feel strongly that it should be included, let me know.

[the remark this answers is not in the document]
Comment (dpw13) [reply doc/hardware/cache/guide.rst:79]: Actually, reading through the docs it appears that you CAN relocate just data. I'll add a small note as this will enable users to relocate buffers without modifying code, but I think I would only recommend this as a last resort.

[the remark this answers is not in the document]
Comment (dpw13) [reply doc/hardware/cache/guide.rst:79]: I keep waffling on this. Thinking more about it, it's also technically possible for users to relocate individual buffers or functions by modifying the linker code, but I wouldn't want to add that to this document. Anybody doing things with that kind of complexity *probably* aren't the target audience for this document. If you still think a reference to `zephyr_code_relocate` should be added to this document, I have the changes typed up and I'll push them, but I'm planning on leaving it out for now.

[the remark this answers is not in the document]
Comment (cfriedt) [reply doc/hardware/cache/guide.rst:79]: Maybe @npitre or @dcpleung or @carlocaione want to comment here (re: documentation on how to put data in uncahed memory regions).

**Review by peter-mitsis — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by kartben — APPROVED** [review]
probably not worth addressing my two minor comments in this PR, but might be worthwhile if a follow-up PR is made at some point.
thanks for this, @dpw13!

File: doc/hardware/arch/arm_cortex_m.rst [hunk 5]
```diff
@@ -469,18 +471,19 @@ For example, to define a new non-cacheable memory region in devicetree:
    };
 
 This will automatically create a new MPU entry in with the correct name, base,
```
[the remark this answers is not in the document]
Comment (dpw13) [reply doc/hardware/arch/arm_cortex_m.rst:473]: Yep, looks like something I wrote... Will follow-up with a quick fix PR.
